### PR TITLE
[Feature:Submission] Late Day database triggers

### PR DESCRIPTION
**Pull request by jportorreal00**
### Please check if the PR fulfills these requirements:

* [ ] Tests for the changes have been added/updated (if possible)
* [ ] Documentation has been updated/added if relevant

### What is the current behavior?
Late days are currently calculated within the LateDays object via php. Current late days use previous late day events as context in order to determine the #of late days remaining and # of late days used.

### What is the new behavior?
Late day information will now be stored in a table. (Table is similar to Late day usage table). This will make accessing late day information a lot faster since we will not have to calculate it from scratch every time.

Late days will now be automatically reset and recalculated using triggers within the databases. Cases where this will happen:

- submit new gradeable
- gradeable version change
- drop submission
- late day (allowed table) addition
- late day (allowed table) deletion
- gradeable late days allowed change
- gradeable due date change
- gradeable due date disabling/enabling
- gradeable late days disabling/enabling
- gradeable deletion
- add new late day extension
- edit existing late day extension
- delete late day extension
- change default late days (this is the only **manual** table reset: inside the php code)

### Other information?

I tested locally on two courses. I ran over each of the scenarios listed and they work as expected (using pgAdmin)

The changes to twig files were due to the changes in the Late Day Usage data. Ex:
image

### CODE REVIEW

File: migration/migrator/migrations/course/20211117160515_add_late_day_triggers.py
Comment (williamjallen) [inline migration/migrator/migrations/course/20211117160515_add_late_day_triggers.py:32]: Is there a particular reason why this doesn't have a timezone?

File: site/app/controllers/admin/ConfigurationController.php
Comment (williamjallen) [inline site/app/controllers/admin/ConfigurationController.php:210]: It would be better to use `===` here instead of `==`

File: site/app/libraries/database/DatabaseQueries.php
Comment (williamjallen) [inline site/app/libraries/database/DatabaseQueries.php:1445]: Is the intention here to use pre- or post-increment for `$index` here?

File: site/app/libraries/database/DatabaseQueries.php
Comment (williamjallen) [inline site/app/libraries/database/DatabaseQueries.php:1428]: It would be good to add type hints to the parameters and return value for this function

File: site/app/libraries/database/DatabaseQueries.php
Comment (williamjallen) [inline site/app/libraries/database/DatabaseQueries.php:1410]: It would be good to add type hints for the return value here

File: site/app/libraries/database/DatabaseQueries.php
Comment (williamjallen) [inline site/app/libraries/database/DatabaseQueries.php:1451]: It would be good to add type hints to the parameters and return value for this function

File: site/app/libraries/database/DatabaseQueries.php
Comment (williamjallen) [inline site/app/libraries/database/DatabaseQueries.php:1476]: It would be good to add type hints to the return value for this function

File: site/app/libraries/database/DatabaseQueries.php
Comment (williamjallen) [inline site/app/libraries/database/DatabaseQueries.php:1492]: Same comment here about type-hinting

File: site/app/models/gradeable/LateDayInfo.php
Comment (williamjallen) [inline site/app/models/gradeable/LateDayInfo.php:107]: Same comment here about type-hinting (applies to all of the following functions)

**Review by williamjallen — CHANGES_REQUESTED** [review]

File: migration/migrator/migrations/course/20211117160515_add_late_day_triggers.py
Comment (jportorreal00) [reply migration/migrator/migrations/course/20211117160515_add_late_day_triggers.py:32]: good point

File: site/app/libraries/database/DatabaseQueries.php
Comment (williamjallen) [reply site/app/libraries/database/DatabaseQueries.php:1483]: Hmm, if I'm understanding what this code does properly, it seems *very* prone to someone making changes in the future and accidentally breaking things.  I would recommend doing something like the following just for clarity.
```
$late_day_events[$index] = $row;
$index++; 
```

File: site/app/libraries/database/DatabaseQueries.php
Comment (williamjallen) [reply site/app/libraries/database/DatabaseQueries.php:1466]: Perhaps Matt can correct me if I'm wrong here but in my opinion, it would be better to use [this](https://www.php.net/manual/en/language.types.declarations.php) type of type hinting.  Using the above method instead of comments, any parameters that are not of the proper type will result in PHP errors in the development environment which helps to catch bugs earlier in my experience.

Comment (jportorreal00) [reply site/app/libraries/database/DatabaseQueries.php:1466]: I see what you mean. In that case do you think I should remove the docstring or keep it?

Comment (williamjallen) [reply site/app/libraries/database/DatabaseQueries.php:1466]: For simple things where the docstring duplicates the information provided by the type hint, I would suggest removing it because it simply clutters the code.  For things where you have specified the return type in more detail, I would argue that it would be good to keep it so that someone can quickly reference it in the future should they wander into this part of the code and wonder what that function returns.  

Matt may have a better suggestion here though.

Comment (jportorreal00) [reply site/app/libraries/database/DatabaseQueries.php:1466]: I see some other functions/constructors with the duplicate type hinting, so I will keep it for now. Thanks!

**Review by DefinitiveAbove — APPROVED** [review]

File: site/app/models/gradeable/LateDayInfo.php
Comment (williamjallen) [inline site/app/models/gradeable/LateDayInfo.php:297]: It would be better to use the the null coalescing (`??`) operator for this I believe.

File: site/tests/app/models/gradeable/LateDaysTester.php
Comment (williamjallen) [inline site/tests/app/models/gradeable/LateDaysTester.php:108]: It would be good to add type hinting here

**Review by williamjallen — CHANGES_REQUESTED** [review]
Code looks good in general, but I think that it would be good to add type hinting in a bunch of places to ensure (at least in the development environment) that errors are thrown if unexpected parameters/return types are provided.
Example with type hinting on two parameters and the return type:
```
public function foo(string param1, int param2): array {
    ...
    return [1, 2, 3];
}
```

File: site/app/models/gradeable/LateDayInfo.php
Comment (jportorreal00) [reply site/app/models/gradeable/LateDayInfo.php]: Since I'm checking $graded_gradeable I cant use the coalescing operator. It might be null, I wont be sure if I can use `getGradeInquiryCount()`. I can potentially shorten this with `$this->graded_gradeable?->getGradeInquiryCount() ?? null;` but we would need PHP v8 and up tp use the `?->` operator.

File: site/app/models/gradeable/LateDayInfo.php
Comment (williamjallen) [inline site/app/models/gradeable/LateDayInfo.php:274]: If I was reading through the code and encountered `getLateDaysCharged()` I would not immediately expect this function to return a negative number.  Rather, I would expect it to return a positive number representing the number of late days charged.  Similarly, it seems rather confusing to have two functions which return the exact same thing.  Is there a reason for this or should `getLateDaysCharged()` be eliminated altogether?

Just to be clear, this isn't a requested change--just food for thought in case changes are desired.

Comment (jportorreal00) [reply site/app/models/gradeable/LateDayInfo.php]: Very good point. I opted to max out the value at 0. I think it could provide better functionality for future changes. I think only late gradeables should return a significant value.

**Review by williamjallen — APPROVED** [review]

File: site/app/templates/grading/electronic/StudentInformationPanel.twig
Comment (williamjallen) [inline site/app/templates/grading/electronic/StudentInformationPanel.twig:90]: It would be better to use `===` here

Comment (jportorreal00) [reply site/app/templates/grading/electronic/StudentInformationPanel.twig:90]: oops this wasnt meant to be in this branch

**Comment by williamjallen** [comment]
I like the most recent changes, but I wonder if this PR is the best PR to include them in.  Would it be better to hold off on making changes to the TA grading interface given that this PR is mostly setting up the triggers?  The new changes could be made as a new PR, or added to one of the other open PRs--I'm not sure which would be best in this case.

**Comment by jportorreal00** [comment]
> I like the most recent changes, but I wonder if this PR is the best PR to include them in. Would it be better to hold off on making changes to the TA grading interface given that this PR is mostly setting up the triggers? The new changes could be made as a new PR, or added to one of the other open PRs--I'm not sure which would be best in this case.

Yes definitely!! It was meant to be part of #7702, but i goofed up the commits somehow. Sorry about that.

**Comment by MasterOdin** [comment]
@jportorreal00 Please try to avoid doing force pushes on reviewed PRs that completely change the history. It makes it hard to know on what commits the reviews were made, and if you made a ton of changes after a review.

**Comment by jportorreal00** [comment]
> @jportorreal00 Please try to avoid doing force pushes on reviewed PRs that completely change the history. It makes it hard to know on what commits the reviews were made, and if you made a ton of changes after a review.

Oh shoot sorry about that! I was pulling from master and didn’t want to deal with merge conflicts. I’ll avoid doing that from now on.

**Comment by MasterOdin** [comment]
Meant to add, do you know what (if any) commits you added after @williamjallen's approval?

**Comment by jportorreal00** [comment]
He approved on 3/2. I added a commit on 3/3 that belonged to #7702. I realized it was wrong so I reverted to the previous commit. So the current state is the commits that William had the chance to look over

**Comment by jportorreal00** [comment]
@MasterOdin heres my local git log

image

**Comment by jportorreal00** [comment]
@williamjallen @MasterOdin There was a new commit added to fix a bug with the Reports for CSVs and rainbow grades. Only electronic gradeables with late days allowed (and not in the future) support late day cache. Any other gradeables don't effect late days, there for they don't have any late day info.

**Pull request merged**
